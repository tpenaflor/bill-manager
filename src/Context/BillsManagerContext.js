import React from 'react'

const BillsManagerContext = React.createContext()

export default BillsManagerContext