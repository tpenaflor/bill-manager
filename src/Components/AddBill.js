import React, {useState} from 'react'
import {useBillsManagerContext} from '../Context/BillsManagerContext'
import { v4 } from 'uuid';

const AddBill = () => {
    const {billDispatch} = useBillsManagerContext()
    const [amount, setAmount] = useState('')
    const [unitCount, setUnitCount] = useState('')
    const [desc, setDesc] = useState('')
    const [splitOpt, setSplitOpt] = useState('even')

    const updateAmount = (e)=> {
      const amnt = e.target.value.trim()
      if (!amnt || amnt.match(/^\d{1,}(\.\d{0,2})?$/))
          return setAmount(amnt)
    }

    const updateUnitCount = (e)=> {
      const count = e.target.value.trim()
      if (!count || count.match(/^\d{1,}?$/))
          return setUnitCount(count)
    }

    const addBill = (e) => {
        e.preventDefault()
        
        if (amount && desc) {
          const id = v4()
          billDispatch({
            type:"ADD_BILL", 
            bill: {
              id,
              desc,
              allocation : {
                participant : [],
                unitCount
              },
              amount,
              splitOpt
            }})

          setDesc('')
          setAmount('')
        }
      }

  return (
    <form onSubmit={addBill}>
      <label>Split Option</label>
      <select 
        id='splitOption' 
        value={splitOpt} 
        onChange={(e)=> { setSplitOpt(e.target.value)}}
      >
        <option value='even'>Split Evenly</option>  
        <option value='unit'>Split per Unit</option>
      </select>

      <input value ={desc} 
        onChange={(e)=>setDesc(e.target.value.trim())}
        placeholder='Bill Description'  />
      {splitOpt === 'unit' && <input value ={unitCount} 
        onChange={updateUnitCount}
        placeholder='Number of Units'  />}
      <input value ={amount} 
        onChange={updateAmount}
        placeholder='Amount'  />
      <button>Add Bill</button>
    </form>
  )
}

export default AddBill