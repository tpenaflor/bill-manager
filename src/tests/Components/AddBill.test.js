import React from 'react'
import {shallow} from 'enzyme'
import { shallowToJson  } from 'enzyme-to-json';

import AddBill from '../../Components/AddBill'
import * as BillsManagerContext from '../../Context/BillsManagerContext'

let billDispatch, wrapper

beforeEach( ()=> {
    billDispatch = jest.fn()
    const contextValues = {
        billDispatch
    }

    jest.spyOn(BillsManagerContext, 'useBillsManagerContext')
        .mockImplementation(() => contextValues)

    wrapper = shallow(<AddBill />)
})

afterEach(()=>{
    expect(shallowToJson(wrapper)).toMatchSnapshot()
})

test('render add bill evenly', ()=> {
    wrapper.find('select').simulate('change', {target:{value: 'even'}})
    expect(shallowToJson(wrapper)).toMatchSnapshot()

    const desc = "test description"
    const amount = "1000"

    wrapper.find('input').at(0).simulate('change', {target:{value: desc}})
    wrapper.find('input').at(1).simulate('change', {target:{value: amount}})

    wrapper.find('form').simulate('submit', {preventDefault:()=>{}})

    expect(billDispatch).toHaveBeenLastCalledWith({
        type:"ADD_BILL", 
        bill: {
          id : expect.any(String),
          desc,
          allocation : {
            participant : [],
            unitCount : ''
          },
          amount,
          splitOpt : 'even'
        }})

})

test('render add bill by unit', ()=> {
    wrapper.find('select').simulate('change', {target:{value: 'unit'}})
    expect(shallowToJson(wrapper)).toMatchSnapshot()

    const desc = "test description"
    const amount = "1000"
    const unitCount = "10"

    wrapper.find('input').at(0).simulate('change', {target:{value: desc}})
    wrapper.find('input').at(1).simulate('change', {target:{value: unitCount}})
    wrapper.find('input').at(2).simulate('change', {target:{value: amount}})

    wrapper.find('form').simulate('submit', {preventDefault:()=>{}})

    expect(billDispatch).toHaveBeenLastCalledWith({
        type:"ADD_BILL", 
        bill: {
          id : expect.any(String),
          desc,
          allocation : {
            participant : [],
            unitCount
          },
          amount,
          splitOpt : 'unit'
        }})

})