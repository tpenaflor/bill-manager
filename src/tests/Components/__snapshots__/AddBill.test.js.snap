// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render add bill by unit 1`] = `
<form
  onSubmit={[Function]}
>
  <label>
    Split Option
  </label>
  <select
    id="splitOption"
    onChange={[Function]}
    value="unit"
  >
    <option
      value="even"
    >
      Split Evenly
    </option>
    <option
      value="unit"
    >
      Split per Unit
    </option>
  </select>
  <input
    onChange={[Function]}
    placeholder="Bill Description"
    value=""
  />
  <input
    onChange={[Function]}
    placeholder="Number of Units"
    value=""
  />
  <input
    onChange={[Function]}
    placeholder="Amount"
    value=""
  />
  <button>
    Add Bill
  </button>
</form>
`;

exports[`render add bill by unit 2`] = `
<form
  onSubmit={[Function]}
>
  <label>
    Split Option
  </label>
  <select
    id="splitOption"
    onChange={[Function]}
    value="unit"
  >
    <option
      value="even"
    >
      Split Evenly
    </option>
    <option
      value="unit"
    >
      Split per Unit
    </option>
  </select>
  <input
    onChange={[Function]}
    placeholder="Bill Description"
    value=""
  />
  <input
    onChange={[Function]}
    placeholder="Number of Units"
    value="10"
  />
  <input
    onChange={[Function]}
    placeholder="Amount"
    value=""
  />
  <button>
    Add Bill
  </button>
</form>
`;

exports[`render add bill evenly 1`] = `
<form
  onSubmit={[Function]}
>
  <label>
    Split Option
  </label>
  <select
    id="splitOption"
    onChange={[Function]}
    value="even"
  >
    <option
      value="even"
    >
      Split Evenly
    </option>
    <option
      value="unit"
    >
      Split per Unit
    </option>
  </select>
  <input
    onChange={[Function]}
    placeholder="Bill Description"
    value=""
  />
  <input
    onChange={[Function]}
    placeholder="Amount"
    value=""
  />
  <button>
    Add Bill
  </button>
</form>
`;

exports[`render add bill evenly 2`] = `
<form
  onSubmit={[Function]}
>
  <label>
    Split Option
  </label>
  <select
    id="splitOption"
    onChange={[Function]}
    value="even"
  >
    <option
      value="even"
    >
      Split Evenly
    </option>
    <option
      value="unit"
    >
      Split per Unit
    </option>
  </select>
  <input
    onChange={[Function]}
    placeholder="Bill Description"
    value=""
  />
  <input
    onChange={[Function]}
    placeholder="Amount"
    value=""
  />
  <button>
    Add Bill
  </button>
</form>
`;
